subxt/src/storage/storage_key.rs

```diff
@@ -0,0 +1,288 @@
+use super::utils::strip_storage_hash_bytes;
+use crate::{
+    dynamic::DecodedValueThunk,
+    error::{Error, StorageAddressError},
+    metadata::{DecodeWithMetadata, Metadata},
+    utils::{Encoded, Static},
+};
+use scale_encode::EncodeAsType;
+use subxt_metadata::StorageHasher;
+
+use derivative::Derivative;
+
+/// This trait should be implemented by anything that can be used as one or multiple storage keys.
+pub trait StorageKey {
+    /// Iterator over the storage keys, each key implements EncodeAsType to
+    /// give the corresponding bytes to a `StorageHasher`.
+    fn keys_iter(&self) -> impl Iterator<Item = &dyn EncodeAsType>;
+
+    /// How many keys are there in total? Each key is an element that needs to be individually hashed.
+    // Note: Ideally we would use `impl ExactSizeIterator<Item = &dyn EncodeAsType>` for `keys_iter` above,
+    // But that plays poorly with the `Flatten` and `Chain` structs.
+    fn keys_len(&self) -> usize;
+
+    /// Attempts to decode the StorageKey from a storage address that has been stripped of its root bytes.
+    ///
+    /// Example: Imagine The `StorageKey` is a tuple (A,B) and the hashers are: [Blake2_128Concat, Twox64Concat].
+    /// Then the memory layout of the storage address is:
+    /// ```txt
+    /// | 8 bytes pallet hash | 8 bytes entry hash | 16 bytes hash of A | ... bytes of A | 8 bytes hash of B | ... bytes of B |
+    /// ```
+    /// `cursor` should point into a region after those first 16 bytes, at the start of a new hash.
+    /// `hashers_and_ty_ids` should consume all the hashers that have been used for decoding, such that there are less hashers coming to the next key.
+    fn decode_from_bytes(
+        cursor: &mut &[u8],
```

Comment:
My gut feeling is that `cursor` should be `&[u8]`, not `&mut &[u8]`; ie this function is just given some bytes to decoee, and some hashers that they represent, and there's no need for it to mutate any external state.

Reply:
I need to think abotu this a bit more because `hashers_and_ty_ids` is in a similar place for me actually.

One advantage of taking `&mut` things is that each `StorageKey` impl can consume the hasher (_and bytes_) that it actually uses up, and hand the rest on to the next `StorageKey` impl (assuming we have eg a tuple of impls). This would let eg `(A, (B, C), D)` to work, because `(B, C)` would consume 2 lots of bytes/hashers when it's called.

However, looking at the code below it doesn't look like we increment the bytes cursor in the same way that we implement the hashers cursor.

This leads to the downside of `&mut` bytes/hashers: to benefit, each implementation _must_ then consume the correct bytes/hasher amount (or error), which is trickier to ensure.

Reply:
I think I had `&[u8]` at first, but if there is a slice of bytes that represents multiple levels of storage keys in a row, we don't know which bytes to hand to the second  storage key, because the offset of where the bytes of the second key starts is not clear until the first key is already decoded. Either we return this offset as a `usize` or use `&mut &[u8]` to chop the bytes off during decoding. But it can be hard to gurantee that the `&mut &[u8]` and the `&mut &[(StorageHasher, u32)]` are in sync.

Reply:
I'm happy sticking with the `&mut` approach for now, having spent more time looking at things :)